### PR TITLE
Update API types for BetterDiscord v1.11.0 (and some other minor changes)

### PR DESCRIPTION
**Pull request by Neodymium7**
The major change here is adding the new API features from the latest BD update, including:
- `Logger` and corresponding bound variant on `BdApi`
- Types for the new exported `Components`and settings APIs
- Types for `showChangelogModal`
- Types for other new utility functions

I also fixed up a few other minor things:
- Removed some unused type declarations in `bdapi.d.ts`
- Renamed `net.ts` to `net.d.ts` to match the other files
- Added more specific types to the `AddonAPI` instances for plugins and themes
- Updated the Tooltip component to include types for props
- Updated `Tooltip` in UI with more specific types

Finally, I updated the `Webpack` typings with generics on the module fetching methods to make them a bit more restrictive.

With the previous behavior, this example's return value would be `any` (which I thought was a little too permissive to be the default given that you can't always know what a query will return), but now it would be `unknown`:

```ts
const foo = BdApi.Webpack.getModule(someFilter);
// Previously - foo: any
// Now - foo: unknown
```

You can also just revert to the previous behavior with:

```ts
const foo = BdApi.Webpack.getModule<any>(someFilter);
```

As a result, I also changed `getWithKey` to return a unique type for the key instead of a string, with an object indexed by that unique type. This way the returned module can only be indexed by a returned key, since the module is only guaranteed to have the returned key anyway. For example:

```ts
const [module, key] = BdApi.Webpack.getWithKey(someFilter);
console.log(module[key]); // No error
console.log(module.someArbitraryKey); // Error!
```

Again, you can revert to the previous behavior with something like:

```ts
const [module, key]: [any, string] = BdApi.Webpack.getWithKey(someFilter);
console.log(module[key]); // No error
console.log(module.someArbitraryKey); // No error
```

Those last few changes are pretty subjective, but I figured it would be better to default to something more restrictive and let the user of the API decide whether or not to use `any` for the returned modules. If you want me to revert those `Webpack` changes before you merge, just let me know :D

### CODE REVIEW

File: addonapi.d.ts [hunk 1]
```diff
@@ -1,4 +1,25 @@
...
+}
+
+export interface PluginInfo extends AddonInfo {
+    exports: Plugin | PluginClass | PluginCallback;
+    instance: any;
+}
+
+export interface ThemeInfo extends AddonInfo {
+    css: string;
+}
```
Comment (Zerthox) [inline addonapi.d.ts]: Not sure if it changed by now, but the object returned by AddonAPI used to be considered an implementation detail and was not documented in the BetterDiscord API.  I am not sure if we want to type BetterDiscord internal objects.

File: components.d.ts [hunk 2]
```diff
@@ -1,3 +1,346 @@
...
+    DropdownInput: React.FunctionComponent<DropdownInputProps>;
+    SettingItem: React.FunctionComponent<SettingItemProps>;
+    KeybindInput: React.FunctionComponent<KeybindInputProps>;
+    NumberInput: React.FunctionComponent<NumberInputProps>;
+    RadioInput: React.FunctionComponent<RadioInputProps>;
+    SearchInput: React.FunctionComponent<SearchInputProps>;
+    SliderInput: React.FunctionComponent<SliderInputProps>;
+    SwitchInput: React.FunctionComponent<SwitchInputProps>;
+    TextInput: React.FunctionComponent<TextInputProps>;
+    SettingGroup: React.FunctionComponent<SettingGroupProps>;
```
Comment (Zerthox) [inline components.d.ts]: Are these guaranteed to be Function/Class components respectively? If not, I think its better to use a generic React Component type. Same thing goes for the Discord Tooltip props, unless BD gives stability guarantees regarding the interface I dont think it should be typed here.

File: contextmenu.d.ts [hunk 3]
```diff
@@ -73,11 +73,11 @@ export interface ContextMenuConfig {
...
-    ControlItem: React.ComponentType<any>;
-    RadioItem: React.ComponentType<any>;
+    Menu: React.FunctionComponent<any>;
+    Group: React.FunctionComponent<any>;
+    Item: React.FunctionComponent<any>;
+    Separator: React.FunctionComponent<{}>;
+    CheckboxItem: React.FunctionComponent<any>;
+    ControlItem: React.FunctionComponent<any>;
+    RadioItem: React.FunctionComponent<any>;
 }
```
Comment (Zerthox) [inline contextmenu.d.ts]: Same thing here, are these guaranteed to be function components? If people need to do Function Component specific stuff with it, I think its better if they have to cast it.

**Comment by Zerthox** [comment]
New APIs and Webpack changes look good. 👍 

Not sure if the return type of `UI.createTooltip()` is considered an implementation detail, the class itself is not documented in BD docs but explicitly named as `Tooltip` rather than `object` like `AddonAPI.get()`. I think its fine to leave the details in for now.

**Review by Zerthox — CHANGES_REQUESTED** [review]

File: addonapi.d.ts (hunk 1 above)
Comment (Neodymium7) [reply addonapi.d.ts]: I don't believe anything's changed, I just wasn't aware of that detail :P Just went ahead and removed it!

File: components.d.ts (hunk 2 above)
Comment (Neodymium7) [reply components.d.ts]: Good catch! Changed everything to just be generic component types and removed Discord implementation-dependent stuff.

File: contextmenu.d.ts (hunk 3 above)
Comment (Neodymium7) [reply contextmenu.d.ts]: Was experimenting with some stuff here before but forgot to revert those type changes as well. I reverted it back to how it was originally.

**Comment by Neodymium7** [comment]
Fixed those issues according to your comments. Let me know if there's anything else I should change or anything I missed!

**Review by Zerthox — APPROVED** [review]

**Comment by Zerthox** [comment]
Thanks!

**Pull request merged**